### PR TITLE
Bring Mailers from Spree and make order shipped email translatable

### LINKED ISSUE #5686: [ByeBye Spree] Move Mailers to OFN
### What we should change and why (this is tech debt)
Move all Mailers to OFN and merge with decorators.

These files:
- core/app/mailers/spree/base_mailer.rb MERGE WITH DECORATOR
- core/app/mailers/spree/order_mailer.rb MERGE WITH DECORATOR
- core/spec/mailers/order_mailer_spec.rb MERGE WITH DECORATOR
- core/app/mailers/spree/shipment_mailer.rb COPY (and improve!)
- core/app/mailers/spree/test_mailer.rb COPY
- core/app/views/spree/shipment_mailer/shipped_email.text.erb COPY
- core/app/views/spree/test_mailer/test_email.text.erb COPY
- core/spec/mailers/shipment_mailer_spec.rb COPY
- core/spec/mailers/test_mailer_spec.rb COPY

### Context
Part of #4826

### Impact and timeline
This will enable us to stop working on top of decorators and refactor the Spree code to a better version (our coding standards are better)

### PR DESCRIPTION
**Pull request by luisramos0**
#### What? Why?

Closes #5686
and addresses point 2 in #3243

Move all needed mailers from spree_core to OFN.

This was we can now translate test email and order shipped email (this email still needs a better template, but that will be for next time).

#### What should we test?
Here we change the base_mailer so better verify email basics like order confirmation and order cancellation for example. Are they sent correctly? Is the data in the email correct? Is the invoice email sent correctly?
Test the emails that were changed:
- test email - is this email sent when the button pressed in the backoffice mail settings?
- shipped email - is this email sent when order is marked as shipped? Any errors in the old un-styled template?

#### Release notes
Changelog Category: Changed
Brought test and shipment emails from Spree. Order Shipped email is now translatable!

### CODE REVIEW

File: app/views/spree/shipment_mailer/shipped_email.html.haml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  = item.variant.product.name
+  = item.variant.options_text
+= "============================================================"
+
+- if @shipment.tracking
+  = Spree.t('shipment_mailer.shipped_email.track_information', tracking: @shipment.tracking)
+- if @shipment.tracking_url
+  = Spree.t('shipment_mailer.shipped_email.track_link', url: @shipment.tracking_url)
+
+= Spree.t('shipment_mailer.shipped_email.thanks')
```
Comment (sauloperez) [inline app/views/spree/shipment_mailer/shipped_email.html.haml]: Are you sure switching from .text to .HTML won't break things? I never investigated this but I think .text is used as a fallback text-version of the email for specific old mail servers/clients. Again, no idea about how it works.

Comment (luisramos0) [reply app/views/spree/shipment_mailer/shipped_email.html.haml]: Spree emails are all text, and ofn's are html. So I think it's fine, from now on these will also be sent in html. Testing will validate that as well.

I think you are referring to something we _could_ do which is to send html and text emails and the email client would use text as a fall back option. This is also good for deliverability but it's something we are not doing and is out of scope for this PR. Makes sense?

Comment (sauloperez) [reply app/views/spree/shipment_mailer/shipped_email.html.haml]: Understood, I thought those text ones were just a fallback and not Spree's only templates :+1: Then, this means that those only-text emails saw Eugeni some time ago were a bug or they were missing an OFN template :see_no_evil:

Comment (luisramos0) [reply app/views/spree/shipment_mailer/shipped_email.html.haml]: The emails seen by Eugeni are this shipment email. It was text in spree and after this PR it's html in OFN. But still without the normal OFN styling. That will need to be addressed in another PR.

**Comment by RachL** [comment]
OMG I'm just discovering this PR!!! Being able to translate this email is such a GOOOD news!!

**Comment by luisramos0** [comment]
I know!! :-)

**Comment by filipefurtad0** [comment]
Hi @luisramos0,

Triggered the following emails, before and after the PR:

Test
Order shipped
Confirm Account 
Welcome to OFN
Reset Passwork
Order Confirmation
Order Cancellation



- Test emails

Before

image


After

image




- Order shipped Emails


Before


image


After

image



- Confirm Account Email - OK, before and after

- Welcome to OFN Email - OK, before and after

- Reset Passwork Email - OK, before and after 



- Order Confirmation emails

These seem to be broken after this PR. I checked this both by clicking "Resend" or "Resend Confirmation" in the backoffice, and in both in this case, the user sees:

image


After placing a new order, neither the Hub nor the customer receive confirmation email. This is a silent error, no snail seen. Check Bugsnag errors for details:

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/5f228433c37fb80018f09297


- Order cancellation emails are also broken, after this PR

image



The changes introduced by the PR work fine, but order confirmation/cancellation emails are quite a critical, so perhaps best not to introduce these in the code, right? I'm moving the PR back to In Dev for a second look.



Also, a product question maybe @RachL @lin-d-hop : I'm not sure what's the difference between the two buttons (pic below). "Resend Confirmation" triggers the modal asking for confirmation but other than that, is there a difference in function?

image

**Comment by RachL** [comment]
> Also, a product question maybe @RachL @lin-d-hop : I'm not sure what's the difference between the two buttons (pic below). "Resend Confirmation" triggers the modal asking for confirmation but other than that, is there a difference in function?

Yes I've always wondered as well. 🤷

**Comment by filipefurtad0** [comment]
In that case, maybe we could consider removing one of them, maybe as a papercut - if appropriate.

The "Resend Confirmation" button has a more descriptive naming and already has the associated modal, which is also helpful I think, so we could keep this one. Thoughts?

**Comment by RachL** [comment]
Sounds good :+1: Let's re-test maybe just to see if it is really 100% the same? Maybe you already did?

**Comment by filipefurtad0** [comment]
Yes, I found no difference - the emails looked identical, but I'd wait for feedback from core Devs as well. Perhaps @luisramos0 and/or reviewers can confirm that the code behind these features is the same?

**Comment by luisramos0** [comment]
yes, the code is the same behind both resends :+1:

Thanks for testing Filipe, I cant replicate that error locally. It's a tricky one related to class loading I think because the "missing method" is there and included as part of one of the helper classes... it should work.

I'll have to come back to this later.

**Comment by filipefurtad0** [comment]
Thanks for your feedback @luisramos0.

The snail-errors were reproducible in staging environments (UK and FR). But thinking of it now, I am not entirely sure the PR broke the order confirmation emails as well, or they just got queued after the snail errors. Maybe testing in a different order would result in a different outcome. I'll try this and provide some additional feedback.

Thanks as well for confirming the code on the "Resend" button :+1: 
Opened an Enhancement/Papercut #5850 on this regard.

**Comment by filipefurtad0** [comment]
I did a quick double-checked now:
Order confirmation emails trigger a bugsnag [error](https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/5f228433c37fb80018f09297?filters[event.since][0]=30d&filters[error.status][0]=open) as well, and don't arrive the mailbox. This happens also after placing an order immediately after staging the PR, so it appears independent of those errors seen by the user (order cancelling and resend confirmation).

**Comment by luisramos0** [comment]
ok, this was easier than I thought. Yes, there was a bug in the code. There's Spree::CheckoutHelper in Spree that was hidden our CheckoutHelper. I managed to replicate the problem you describe, I pushed a new simple commit (no need for code review) and I restaged the PR. The problem is now gone. This is ready for testing again :+1:

**Comment by luisramos0** [comment]
In staging I got the shipment email like this:
image

So I added some paragraphs and breaklines to the template so it doesnt happen again.
We need to revisit this email template. I think we need a new issue to redo the shipment email template, like this one: https://github.com/openfoodfoundation/openfoodnetwork/issues/2910

**Comment by luisramos0** [comment]
ok, the email now looks like this :+1:

image

**Comment by RachL** [comment]
I've re-tested them all:

- Test : I'm not sure how to get the email. Should I turn one of my accounts as super admin? 
- Order shipped ✅ 
- Confirm Account ✅ 
- Welcome to OFN ✅ 
- Reset Password ✅ 
- Send Invoice ✅ 
- Order Confirmation ✅ 
- Order Cancellation ✅ 

As all email works perfectly, I'm moving this to ready to go!

**Comment by RachL** [comment]
Nevermind: I found how to test the test email :)

image

**Pull request merged**